### PR TITLE
[13.0][ADD] account invoice customer no autofollow

### PR DESCRIPTION
**Pull request by dessanhemrayev**
By default Odoo adds customer as follower to all invoices. This module allows to disable this behaviour using General Settings.

### CODE REVIEW

File: account_invoice_partner_no_autofollow/models/account_move.py
Comment (ivs-cetmix) [inline account_invoice_partner_no_autofollow/models/account_move.py]: it's not quite "override" rather "result" or "res"

File: account_invoice_partner_no_autofollow/models/account_move.py
Comment (ivs-cetmix) [inline account_invoice_partner_no_autofollow/models/account_move.py]: This is not the right way to do this.
Use "unsubscribe follower" method of mail thread instead.

File: account_invoice_partner_no_autofollow/models/account_move.py
Comment (ivs-cetmix) [inline account_invoice_partner_no_autofollow/models/account_move.py]: Same as mentioned above here

**Review by ivs-cetmix — CHANGES_REQUESTED** [review]

**Comment by yostashiro** [comment]
@dessanhemrayev I think what you want to do can be achieved with https://github.com/OCA/social/tree/13.0/mail_restrict_follower_selection. Have you tried this module?

File: account_invoice_partner_no_autofollow/README.rst
Comment (ivs-cetmix) [inline account_invoice_partner_no_autofollow/README.rst]: Leave this file blank. Add module description to /readme instead.
Check here: https://github.com/OCA/maintainer-tools/tree/master/template/module/readme

**Review by ivs-cetmix — CHANGES_REQUESTED** [review]
You are currently trying to remove follower after he/she was added.
However more efficient way will be prevent this follower from bering subscribed at all.
For this it is good to override [message_subscribe](https://github.com/odoo/odoo/blob/55fbbfc26d0568b1b1c53ccf7700b703c8b1bb8b/addons/mail/models/mail_thread.py#L2817) function and perform a regular check based on context passed from action_post

**Comment by ivs-cetmix** [comment]
> @dessanhemrayev I think what you want to do can be achieved with https://github.com/OCA/social/tree/13.0/mail_restrict_follower_selection. Have you tried this module?

hi @yostashiro , we have checked that module of course. The main difference is that our module is used to fix only a single occurrence of the customer subscription. It doesn't imply any further limitations so user can subscribe customer later explicitly if needed.

File: account_invoice_partner_no_autofollow/README.rst
Comment (dessanhemrayev) [reply account_invoice_partner_no_autofollow/README.rst]: Okey)

File: account_invoice_partner_no_autofollow/tests/test_account_move_no_autofollow.py
Comment (ivs-cetmix) [inline account_invoice_partner_no_autofollow/tests/test_account_move_no_autofollow.py]: Fix spelling: Follower count must be equal **to** 0
Here and in all other message strings.
You are comparing to 1 but message says "must be equal **to** 0"
And why 0? Using fixed numbers here might play a bad trick on you.
What if some other module will add another partners as followers? Eg "accountant assistant" etc?

File: account_invoice_partner_no_autofollow/models/account_move.py
Comment (ivs-cetmix) [inline account_invoice_partner_no_autofollow/models/account_move.py]: Add default value here please:
`get_param("account.no_auto_follow", False)`

File: account_invoice_partner_no_autofollow/readme/DESCRIPTION.rst
Comment (ivs-cetmix) [inline account_invoice_partner_no_autofollow/readme/DESCRIPTION.rst]: By default Odoo adds customer as follower to all invoices.
This module allows to disable this behaviour using General Settings.

File: account_invoice_partner_no_autofollow/__manifest__.py
Comment (ivs-cetmix) [inline account_invoice_partner_no_autofollow/__manifest__.py]: Add maintainers here

File: account_invoice_partner_no_autofollow/tests/test_account_move_no_autofollow.py
Comment (ivs-cetmix) [inline account_invoice_partner_no_autofollow/tests/test_account_move_no_autofollow.py]: Why 1? Using fixed numbers here might play a bad trick on you.
What if some other module will add another partners as followers? Eg "accountant assistant" etc?

File: account_invoice_partner_no_autofollow/tests/test_account_move_no_autofollow.py
Comment (ivs-cetmix) [inline account_invoice_partner_no_autofollow/tests/test_account_move_no_autofollow.py]: You are comparing to 1 but message says "must be equal **to** 0"
And why 0? Using fixed numbers here might play a bad trick on you.
What if some other module will add another partners as followers? Eg "accountant assistant" etc?

File: account_invoice_partner_no_autofollow/views/res_config_settings.xml
Comment (ivs-cetmix) [inline account_invoice_partner_no_autofollow/views/res_config_settings.xml]: Customer will not be added as follower automatically

File: account_invoice_partner_no_autofollow/models/account_move.py
Comment (ivs-cetmix) [inline account_invoice_partner_no_autofollow/models/account_move.py]: Why do unsubscribe partners here explicitly if this can be done via context?

File: account_invoice_partner_no_autofollow/models/account_move.py
Comment (ivs-cetmix) [inline account_invoice_partner_no_autofollow/models/account_move.py]: The goal of the module is to prevent only the invoiced partner from being subscribed.
Here you block **all** partners from subscription.

**Review by ivs-cetmix — CHANGES_REQUESTED** [review]
If a user downloads this module how will he know how to configure it?
You haven't provided any configuration information in the module description.

File: account_invoice_partner_no_autofollow/models/res_config.py
Comment (ivs-cetmix) [inline account_invoice_partner_no_autofollow/models/res_config.py]: rename file to res_config_settings.py
Check the OCA guidelines [here](https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#file-naming)

File: account_invoice_partner_no_autofollow/models/res_config.py
Comment (ivs-cetmix) [inline account_invoice_partner_no_autofollow/models/res_config.py]: string="Customer disable autofollow"

Comment (dessanhemrayev) [reply account_invoice_partner_no_autofollow/models/res_config.py]: ok, done

File: account_invoice_partner_no_autofollow/models/res_config.py
Comment (dessanhemrayev) [reply account_invoice_partner_no_autofollow/models/res_config.py]: done

File: account_invoice_partner_no_autofollow/models/account_move.py
Comment (dessanhemrayev) [reply account_invoice_partner_no_autofollow/models/account_move.py]: fixed

File: account_invoice_partner_no_autofollow/models/account_move.py
Comment (dessanhemrayev) [reply account_invoice_partner_no_autofollow/models/account_move.py]: fixed

File: account_invoice_partner_no_autofollow/views/res_config_settings.xml
Comment (dessanhemrayev) [reply account_invoice_partner_no_autofollow/views/res_config_settings.xml]: done

File: account_invoice_partner_no_autofollow/tests/test_account_move_no_autofollow.py
Comment (dessanhemrayev) [reply account_invoice_partner_no_autofollow/tests/test_account_move_no_autofollow.py]: ok, fixed

File: account_invoice_partner_no_autofollow/__manifest__.py
Comment (dessanhemrayev) [reply account_invoice_partner_no_autofollow/__manifest__.py]: done

File: account_invoice_partner_no_autofollow/readme/DESCRIPTION.rst
Comment (dessanhemrayev) [reply account_invoice_partner_no_autofollow/readme/DESCRIPTION.rst]: done

File: account_invoice_partner_no_autofollow/models/account_move.py
Comment (dessanhemrayev) [reply account_invoice_partner_no_autofollow/models/account_move.py]: ok

File: account_invoice_partner_no_autofollow/tests/test_account_move_no_autofollow.py
Comment (dessanhemrayev) [reply account_invoice_partner_no_autofollow/tests/test_account_move_no_autofollow.py]: fix

File: account_invoice_customer_no_autofollow/models/account_move.py
Comment (ivs-cetmix) [inline account_invoice_customer_no_autofollow/models/account_move.py]: You have no record here yet. Calling anything is pointless.

File: account_invoice_customer_no_autofollow/readme/CONFIGURE.rst
Comment (ivs-cetmix) [inline account_invoice_customer_no_autofollow/readme/CONFIGURE.rst]: Enable “Customer disable autofollow”;

**Review by ivs-cetmix — CHANGES_REQUESTED** [review]

File: account_invoice_customer_no_autofollow/readme/CONFIGURE.rst
Comment (dessanhemrayev) [reply account_invoice_customer_no_autofollow/readme/CONFIGURE.rst]: fixed

File: account_invoice_customer_no_autofollow/models/account_move.py
Comment (dessanhemrayev) [reply account_invoice_customer_no_autofollow/models/account_move.py]: fixed

**Review by ivs-cetmix — APPROVED** [review]
Looks good. Please squash commits and move PR out of draft

**Comment by isserver1** [comment]
@OCA/accounting-maintainers would appreciate someone checking this PR too. So we could merge it.

File: account_invoice_customer_no_autofollow/models/res_config_settings.py
Comment (ivs-cetmix) [inline account_invoice_customer_no_autofollow/models/res_config_settings.py]: field name should be more explicit
eg "invoice_partner_no_autofollow"

File: account_invoice_customer_no_autofollow/readme/CONFIGURE.rst
Comment (ivs-cetmix) [inline account_invoice_customer_no_autofollow/readme/CONFIGURE.rst]: Go to General Settings/Invoicing (or Accounting)

File: account_invoice_customer_no_autofollow/__manifest__.py
Comment (ivs-cetmix) [inline account_invoice_customer_no_autofollow/__manifest__.py]: name should follow CamelCase convention
eg "Invoice Customer No Autofollow"

File: account_invoice_customer_no_autofollow/__manifest__.py
Comment (ivs-cetmix) [inline account_invoice_customer_no_autofollow/__manifest__.py]: summary must contain some valuable summary. 
Eg "Do not add customer as follower in Invoices"

**Review by ivs-cetmix — CHANGES_REQUESTED** [review]
- Check comments
- Configuration setting has two modes: enables and disables. Bot of them must be covered by tests.

File: account_invoice_customer_no_autofollow/models/res_config_settings.py
Comment (dessanhemrayev) [reply account_invoice_customer_no_autofollow/models/res_config_settings.py]: Got it, I'll do it

File: account_invoice_customer_no_autofollow/readme/CONFIGURE.rst
Comment (dessanhemrayev) [reply account_invoice_customer_no_autofollow/readme/CONFIGURE.rst]: Got it, I'll fix it

File: account_invoice_customer_no_autofollow/__manifest__.py
Comment (dessanhemrayev) [reply account_invoice_customer_no_autofollow/__manifest__.py]: Okey, I'll fix it

File: account_invoice_customer_no_autofollow/__manifest__.py
Comment (dessanhemrayev) [reply account_invoice_customer_no_autofollow/__manifest__.py]: Got it, I'll fix it

File: account_invoice_customer_no_autofollow/tests/test_account_move_autofollow_enabled.py
Comment (thoemy) [inline account_invoice_customer_no_autofollow/tests/test_account_move_autofollow_enabled.py]:
```python
            msg="The customer must be among the subscribers",
```
```suggestion
            msg="The customer must be among the subscribers",
```

File: account_invoice_customer_no_autofollow/models/account_move.py
Comment (thoemy) [inline account_invoice_customer_no_autofollow/models/account_move.py:17]:
```suggestion
            if partner_ids and self.partner_id.id in partner_ids:
                partner_ids.remove(self.partner_id.id)
```

I managed to get the following exception here. A test for the case where the partner is not in the list would probably make sense. This seems to be the case when the partner was added as follower before posting the invoice.

```
Odoo Server Error
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 624, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 310, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/usr/lib/python3/dist-packages/odoo/tools/pycompat.py", line 14, in reraise
    raise value
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 669, in dispatch
    result = self._call_function(**self.params)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 350, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/odoo/service/model.py", line 94, in wrapper
    return f(dbname, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 339, in checked_call
    result = self.endpoint(*a, **kw)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 915, in __call__
    return self.method(*args, **kw)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 515, in response_wrap
    response = f(*args, **kw)
  File "/usr/lib/python3/dist-packages/odoo/addons/web/controllers/main.py", line 1343, in call_button
    action = self._call_kw(model, method, args, kwargs)
  File "/usr/lib/python3/dist-packages/odoo/addons/web/controllers/main.py", line 1331, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/usr/lib/python3/dist-packages/odoo/api.py", line 387, in call_kw
    result = _call_kw_multi(method, model, args, kwargs)
  File "/usr/lib/python3/dist-packages/odoo/api.py", line 374, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/odoo/addons_custom/account_invoice_customer_no_autofollow/models/account_move.py", line 31, in action_post
    .get_param("account.no_auto_follow", False)
  File "/usr/lib/python3/dist-packages/odoo/addons/account/models/account_move.py", line 2362, in action_post
    return self.post()
  File "/usr/lib/python3/dist-packages/odoo/addons_custom/sale_commission/models/account_move.py", line 35, in post
    return super().post()
  File "/usr/lib/python3/dist-packages/odoo/addons_ecoservice/ecoservice_financeinterface_datev/models/account/account_move.py", line 67, in post
    return super().post()
  File "/usr/lib/python3/dist-packages/odoo/addons_ecoservice/ecoservice_financeinterface/models/account/account_move.py", line 81, in post
    return super().post()
  File "/usr/lib/python3/dist-packages/odoo/addons/account_asset/models/account_move.py", line 36, in post
    res = super(AccountMove, self).post()
  File "/usr/lib/python3/dist-packages/odoo/addons/sale/models/account_invoice.py", line 81, in post
    res = super(AccountMove, self).post()
  File "/usr/lib/python3/dist-packages/odoo/addons/purchase_stock/models/account_invoice.py", line 182, in post
    return super(AccountMove, self).post()
  File "/usr/lib/python3/dist-packages/odoo/addons/l10n_de/models/datev.py", line 35, in post
    return super(AccountMove, self).post()
  File "/usr/lib/python3/dist-packages/odoo/addons/account_reports/models/account_activity.py", line 83, in post
    return super(AccountMove, self).post()
  File "/usr/lib/python3/dist-packages/odoo/addons/stock_account/models/account_move.py", line 52, in post
    res = super(AccountMove, self).post()
  File "/usr/lib/python3/dist-packages/odoo/addons_ecoservice/ecoservice_partner_account-13/models/account/account_move.py", line 13, in post
    return super(AccountMove, self).post()
  File "/usr/lib/python3/dist-packages/odoo/addons/account_invoice_extract/models/account_invoice.py", line 288, in post
    res = super(AccountMove, self).post()
  File "/usr/lib/python3/dist-packages/odoo/addons/account/models/account_move.py", line 2295, in post
    move.message_subscribe([p.id for p in [move.partner_id] if p not in move.sudo().message_partner_ids])
  File "/usr/lib/python3/dist-packages/odoo/addons_custom/account_invoice_customer_no_autofollow/models/account_move.py", line 9, in message_subscribe
    partner_ids.remove(self.partner_id.id)
ValueError: list.remove(x): x not in list
```

**Review by thoemy — CHANGES_REQUESTED** [review]

File: account_invoice_customer_no_autofollow/models/account_move.py
Comment (dessanhemrayev) [reply account_invoice_customer_no_autofollow/models/account_move.py:17]: Thank you for this hint! Missed it somehow. Would appreciate your review when fixed

**Comment by Aldeigja** [comment]
@OCA/accounting-maintainers could you take a look?

**Review by GabbasovDinar — APPROVED** [review]
Code LGTM! Please squash commits

**Comment by grindtildeath** [comment]
Cannot the same be achieved using https://github.com/OCA/social/tree/13.0/mail_restrict_follower_selection :question: 

EDIT: Sorry I missed that the question was already asked by @yostashiro 

EDIT 2: Shouldn't such a module be added in the social repository and use a generic way to disable auto subscription that is not tied to a particular model though?

**Comment by ivs-cetmix** [comment]
> Cannot the same be achieved using https://github.com/OCA/social/tree/13.0/mail_restrict_follower_selection ❓
> 
> EDIT: Sorry I missed that the question was already asked by @yostashiro
> 
> EDIT 2: Shouldn't such a module be added in the social repository and use a generic way to disable auto subscription that is not tied to a particular model though?

Sounds reasonable. However this is much more related to Account invoice than to Social. Actually this repo will be the first place someone will be looking for such feature. 
Meanwhile we had done a similar module for Sales and it is already merged:
https://github.com/OCA/sale-workflow/tree/13.0/sale_order_partner_no_autofollow

**Comment by Aldeigja** [comment]
@OCA/accounting-maintainers could you review, please?

**Comment by Aldeigja** [comment]
@OCA/accounting-maintainers please review when you have a moment. Thank you!

**Review by ivs-cetmix — APPROVED** [review]
LGTM!

**Comment by ivs-cetmix** [comment]
@thoemy could you check once again pls?

**Review by thoemy — APPROVED** [review]

**Comment by Aldeigja** [comment]
@OCA/accounting-maintainers is it good enough for merge?

**Comment by ivs-cetmix** [comment]
@OCA/accounting-maintainers could you have a look please?

**Comment by dessanhemrayev** [comment]
@OCA/accounting-maintainers could you have a look please?

**Review by grindtildeath — COMMENTED** [review]
Could you please add headers to your python files? (not needed for `__init__.py`)

Would also be good if you can add a development status.

I'll merge it when done.

**Comment by dessanhemrayev** [comment]
@grindtildeath Thanks for review, but how do I add a development status?

**Comment by grindtildeath** [comment]
@dessanhemrayev It's not strictly mandatory, but at least in can ease getting things merged :wink:

https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/oca_module_lifecycle_development_status.rst

**Comment by dessanhemrayev** [comment]
> @dessanhemrayev It's not strictly mandatory, but at least in can ease getting things merged 😉
> 
> https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/oca_module_lifecycle_development_status.rst

@grindtildeath, Thanks, I added it
check please)

**Pull request merged**
